File: day5.py
import collections

with open("day5.txt") as f:
  lines = f.read().rstrip().split("\n")
  split = lines.index("")
  stack_lines = lines[:split]
  instructions = lines[split+1:]

  stacks = collections.defaultdict(list)
  for i, k in enumerate(stack_lines[-1].split()):
    col = i*4 + 1
    for line in reversed(stack_lines[:-1]):
      if line[col] != ' ':
        stacks[k].append(line[col])

  for l in instructions:
    parts = l.split(" ")
    count = int(parts[1])
    src = parts[3]
    to = parts[5]

    stacks[to] += reversed(stacks[src][-count:])  # part 1
    # stacks[to] += stacks[src][-count:]  # part 2
    stacks[src] = stacks[src][:-count]

  print("".join(s[-1] for s in stacks.values()))

File: src/bin/day5.rs
extern crate itertools;

use std::fs::File;
use std::io::{self, BufRead};

fn main() -> io::Result<()> {
    let file = File::open("day5.txt")?;
    let mut reading_stack = true;
    let mut stacks = Vec::<Vec<char>>::new();
    for line in io::BufReader::new(file).lines() {
        let line = line?;

        if line.len() == 0 {
            continue;
        }
        if reading_stack {
            let line: Vec<char> = line.chars().collect();
            let count = (line.len() + 1) / 4;
            if stacks.len() == 0 {
                for _ in 0..count {
                    stacks.push(Vec::new());
                }
            }
            if line[1] == '1' {
                reading_stack = false;
                for s in &mut stacks {
                    s.reverse();
                }
            } else {
                for i in 0..count {
                    let col = i*4 + 1;
                    if line[col] != ' ' {
                        stacks[i].push(line[col]);
                    }
                }
            }
        } else {
            let parts = line.split(' ').collect::<Vec<&str>>();
            let count = parts[1].parse::<usize>().unwrap();
            let src = parts[3].parse::<usize>().unwrap() - 1;
            let dest = parts[5].parse::<usize>().unwrap() - 1;

            // How to write these with extend without fighting the borrow checker?

            // Part 1:
            for _ in 0..count {
                let v = stacks[src].pop().unwrap();
                stacks[dest].push(v);
            }

            // Part 2:
            // let src_keep = stacks[src].len() - count;
            // let mut v = stacks[src].split_off(src_keep);
            // stacks[dest].append(&mut v);
        }
    }
    for s in &stacks {
        print!("{}", s[s.len() - 1]);
    }
    println!();
    Ok(())
}
